On OS(DARWIN), enable ExecutableAllocator to use a constexpr page size instead of calling WTF::pageSize() repeatedly.

MenloDorian opened the pull request:
#### 733bed1be716b30a5630ba3f407094ea94530e19
<pre>
On OS(DARWIN), enable ExecutableAllocator to use a constexpr page size instead of calling WTF::pageSize() repeatedly.
<a href="https://bugs.webkit.org/show_bug.cgi?id=242971">https://bugs.webkit.org/show_bug.cgi?id=242971</a>

Reviewed by Yusuke Suzuki.

1. On OS(DARWIN), the page size is fixed.  We already rely on this in MarkedBlock.
   Allow ExecutableAllocator to do the same and use the constexpr CeilingOnPageSize value as the
   page size instead of repeatedly calling WTF::pageSize().
2. Moved isPowerOfTwo() to StdLibExtra.h, and made it constexpr.
3. Made roundUpToMultipleOf() constexpr.

   The above working together allows many page size calculations to be constant folded.

4. Added a comment to clarify why each allocatable region in the RegionAllocators must be able to
   fit 2 jump islands in a 128M span (for ARM64), and not just one.  This is because the jumps in
   the jump islands can chain using relative jumps to adjacent islands as well.  Hence, any
   adjacent neighbor jump island needs to be within the ARM64 128M near jump range.

5. Removed some unused code.

* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::emitCall):
* Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
(JSC::initializeJITPageReservation):
* Source/JavaScriptCore/jit/ExecutableAllocator.h:
* Source/JavaScriptCore/jit/JITCall.cpp:
(JSC::JIT::compileSetupFrame):
* Source/WTF/wtf/PageBlock.cpp:
(WTF::pageSize):
(WTF::pageMask): Deleted.
* Source/WTF/wtf/PageBlock.h:
(WTF::isPageAligned):
(WTF::isPowerOfTwo): Deleted.
* Source/WTF/wtf/StdLibExtras.h:
(WTF::isPowerOfTwo):
(WTF::roundUpToMultipleOfImpl):
(WTF::roundUpToMultipleOf):

Canonical link: <a href="https://commits.webkit.org/252775@main">https://commits.webkit.org/252775@main</a>
</pre>

saambarati commented on Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
```diff
@@ -446,19 +458,21 @@ class FixedVMPoolExecutableAllocator final {
-            RELEASE_ASSERT(isPageAligned(m_regionSize));
-            const unsigned numAllocators = (reservation.size + m_regionSize - 1) / m_regionSize;
+            const size_t islandRegionSize = roundUpToMultipleOf(EXECUTABLE_PAGE_SIZE, static_cast<size_t>(maxRegionSize * islandRegionSizeFraction));
+            m_regionSize = maxRegionSize - islandRegionSize;
```
I believe this change is wrong for the reason we talked about in slack. But commenting here for posterity.

What this change does is make the layout:
```
[  island 0 ] [ JIT region          ] [ island 1 ]
-16          0                     112           128
```

But it should be:
```
[  island 0 ] [ JIT region          ] [ island 1 ]
0           16                      112          128
```

So any jump in island 0 can make it to island 1 using a near branch.

MenloDorian replied:
Undid the erroneous "fix" and added a comment to document and clarify this detail instead.

Constellation commented on Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
```diff
@@ -86,6 +86,14 @@ namespace JSC {
+// Just use the constexpr CeilingOnPageSize for better efficiency.
+#define EXECUTABLE_PAGE_SIZE CeilingOnPageSize
+#else
+#define EXECUTABLE_PAGE_SIZE WTF::pageSize()
```
I wonder if we should use `executablePageSize()` function instead of macro (since usually we use macro values for constexpr constant, but in this case, one of them are function call).
```cpp
#if OS(DARWIN) && CPU(ARM64)
static inline constexpr size_t executablePageSize() { return CeilingOnPageSize; }
#else
static inline size_t executablePageSize() { return WTF::pageSize(); }
#endif

MenloDorian replied:
Sounds good.  I will apply this.

Constellation reviewed (approved):
r=me with comment.

Commit:
On OS(DARWIN), enable ExecutableAllocator to use a constexpr page size instead of calling WTF::pageSize() repeatedly.
https://bugs.webkit.org/show_bug.cgi?id=242971

Reviewed by Yusuke Suzuki.

1. On OS(DARWIN), the page size is fixed.  We already rely on this in MarkedBlock.
   Allow ExecutableAllocator to do the same and use the constexpr CeilingOnPageSize value as the
   page size instead of repeatedly calling WTF::pageSize().
2. Moved isPowerOfTwo() to StdLibExtra.h, and made it constexpr.
3. Made roundUpToMultipleOf() constexpr.

   The above working together allows many page size calculations to be constant folded.

4. Added a comment to clarify why each allocatable region in the RegionAllocators must be able to
   fit 2 jump islands in a 128M span (for ARM64), and not just one.  This is because the jumps in
   the jump islands can chain using relative jumps to adjacent islands as well.  Hence, any
   adjacent neighbor jump island needs to be within the ARM64 128M near jump range.

5. Removed some unused code.

* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::emitCall):
* Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
(JSC::initializeJITPageReservation):
* Source/JavaScriptCore/jit/ExecutableAllocator.h:
* Source/JavaScriptCore/jit/JITCall.cpp:
(JSC::JIT::compileSetupFrame):
* Source/WTF/wtf/PageBlock.cpp:
(WTF::pageSize):
(WTF::pageMask): Deleted.
* Source/WTF/wtf/PageBlock.h:
(WTF::isPageAligned):
(WTF::isPowerOfTwo): Deleted.
* Source/WTF/wtf/StdLibExtras.h:
(WTF::isPowerOfTwo):
(WTF::roundUpToMultipleOfImpl):
(WTF::roundUpToMultipleOf):

Canonical link: https://commits.webkit.org/252775@main
